src/dataflow/src/render/mod.rs

```diff
@@ -740,6 +740,63 @@ impl<G> Context<G, RelationExpr, Row, Timestamp>
 where
     G: Scope<Timestamp = Timestamp>,
 {
+    fn extract_map_filter_project(
+        &mut self,
+        relation_expr: &RelationExpr,
+    ) -> Option<(MapFilterProject, RelationExpr)> {
+        match relation_expr {
+            RelationExpr::Map { input, scalars } => {
+                let (mfp, input) = self.extract_map_filter_project(input)?;
+                Some((mfp.map(scalars.iter().cloned()), input))
+            }
+            RelationExpr::Filter { input, predicates } => {
+                let (mfp, input) = self.extract_map_filter_project(input)?;
+                Some((mfp.filter(predicates.iter().cloned()), input))
+            }
+            RelationExpr::Project { input, outputs } => {
+                let (mfp, input) = self.extract_map_filter_project(input)?;
+                Some((mfp.project(outputs.iter().cloned()), input))
+            }
+            RelationExpr::Get { .. } => Some((
+                MapFilterProject::new(relation_expr.arity()),
+                relation_expr.clone(),
+            )),
+            _ => None,
+        }
+    }
+
+    /// Attempt to extract a chain of map/filter/project operators on top of a Get. Returns true if
+    /// it was successful and false otherwise. We still fall back to individual implementations for
+    /// the various operators so that ones like Filter, that special-case being on top of a join,
+    /// can still do so.
+    fn try_render_map_filter_project(
+        &mut self,
+        relation_expr: &RelationExpr,
+        scope: &mut G,
+        worker_index: usize,
+    ) -> bool {
+        if let Some((mfp, input)) = self.extract_map_filter_project(relation_expr) {
+            self.ensure_rendered(&input, scope, worker_index);
+            let (ok_collection, mut err_collection) = self.collection(&input).unwrap();
```

Comment:
I think this is the moment where we could integrate this more tightly, using the `flat_map_ref` method instead of `collection`. It has the potential to avoid pulling the contents of the arrangement first, and then `flat_map_fallible`'ing them afterwards, and instead doing all the work at once.